mountpoint-s3/tests/fuse_tests/unlink_test.rs

```diff
@@ -0,0 +1,104 @@
+use std::fs::{self, File};
+use std::io::{ErrorKind, Read, Seek, SeekFrom};
+
+use fuser::BackgroundSession;
+use mountpoint_s3::S3FilesystemConfig;
+use tempfile::TempDir;
+use test_case::test_case;
+
+use crate::fuse_tests::{read_dir_to_entry_names, TestClientBox};
+
+/// Simple test cases, assuming a file isn't open for reading elsewhere.
+fn simple_unlink_tests<F>(creator_fn: F, prefix: &str)
+where
+    F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    let (mount_point, _session, mut test_client) = creator_fn(prefix, Default::default());
+
+    // Add a file directly to the bucket
+    test_client.put_object("dir/hello.txt", b"hello world").unwrap();
+    test_client.put_object("dir/foo.txt", b"bar").unwrap();
+
+    let main_dir = mount_point.path().join("dir");
+    let path = mount_point.path().join("dir/hello.txt");
+    let nonexistent_path = mount_point.path().join("dir/not-here.txt");
+
+    // files should be visible in readdir
+    let read_dir_iter = fs::read_dir(&main_dir).unwrap();
+    let dir_entry_names = read_dir_to_entry_names(read_dir_iter);
+    assert_eq!(dir_entry_names, vec!["foo.txt", "hello.txt"]);
+
+    let err = fs::remove_file(nonexistent_path).expect_err("file remove/unlink on non-existing file should fail");
+    assert_eq!(err.kind(), ErrorKind::NotFound);
+
+    fs::remove_file(&path).expect("file remove/unlink of existing path should succeed");
+
+    let err = fs::remove_file(path).expect_err("file remove/unlink a second time on existing file should fail");
+    assert_eq!(err.kind(), ErrorKind::NotFound);
+
+    // readdir should now show the file as gone
+    let read_dir_iter = fs::read_dir(&main_dir).unwrap();
+    let dir_entry_names = read_dir_to_entry_names(read_dir_iter);
+    assert_eq!(dir_entry_names, vec!["foo.txt"]);
+}
+
+#[cfg(feature = "s3_tests")]
+#[test]
+fn simple_unlink_test_s3() {
+    simple_unlink_tests(crate::fuse_tests::s3_session::new, "simple_unlink_tests");
+}
+
+#[test_case(""; "no prefix")]
+#[test_case("simple_unlink_test"; "prefix")]
+fn simple_unlink_test_mock(prefix: &str) {
+    simple_unlink_tests(crate::fuse_tests::mock_session::new, prefix);
+}
+
+/// Testing behavior when a file is unlinked in the middle of reading
```

Comment:
Should we also test attempts to unlink while writing?

Reply:
Yes we should, I'll add this